### PR TITLE
[Update] Oauth section update

### PR DESCRIPTION
**Pull request by hzoppetti**
Updated the OAuth section to include:
- link to docs
- flow for private vs public
- before you begin info on creating a client
- info on refresh tokens
- links to endpoints

### CODE REVIEW

File: openapi.yaml [hunk 1]
```diff
@@ -47,33 +47,45 @@ info:
     | **HTTP Authorization Scheme** | bearer |
 
     ## OAuth
+    If all you need is a Personal Access Token, feel free to skip ahead to the next section.
 
-    The OAuth workflow is a three-step process to authenticate a User before an
-    application can start making API calls on the User's behalf. If all you need
-    is a Personal Access Token, feel free to skip ahead to the next section.
+    For more help on OAuth see our guide on [How to Create an OAuth App with teh Linode Python API Library](https://www.linode.com/docs/platform/api/how-to-create-an-oauth-app-with-the-linode-python-api-library/#oauth-2-authentication-exchange).
```
Comment (Dorthu) [inline openapi.yaml]: > with teh Linode

File: openapi.yaml [hunk 2]
```diff
@@ -47,33 +47,45 @@ info:
...
 
-    First, the User visits the application's website and is directed to log with
-    Linode. The User is then redirected to Linode's authentication server and
-    presented the scope levels the application is requesting. Once the User
-    accepts the request for access, we redirect them back to the application's
-    specified redirect URI with an access code.
+    Before you begin, you need to [Create an OAuth Client](https://developers.linode.com/api/v4/account-oauth-clients/#post).
```
Comment (Dorthu) [inline openapi.yaml]: Might also consider https://cloud.linode.com/profile/clients

File: openapi.yaml [hunk 3]
```diff
@@ -47,33 +47,45 @@ info:
...
+    | 3.  The user logs into the authorization server with their username and password. | 3.  The user logs into the authorization server with their username and password. |
+    | 4.  The authorization server redirects the user to the specificed redirect URL with a temporary authorization `code` (exchange code) in the URL. | 4.  The authorization server redirects the user back to your applaction with an OAuth token in the URL hash. This is temporary and expires in 2 hours without a `refresh_token`. |
+    | 5.  The application issues a POST request (*see below*) to the authentication server with the exchange code, `client_id`, and the client application's `client_secret`. In return, the server returns an OAuth token and `refresh_token`. |  |
+    | 6.  The authentication server responds to the client application with a new OAuth `access_token`. `access_token` is set to expire in 2 hours if not refreshed. |  |
```
Comment (Dorthu) [inline openapi.yaml]: Actually, the token will just expire in 2 hours.  If refreshed, a new token is generated.

File: openapi.yaml [hunk 4]
```diff
@@ -47,33 +47,45 @@ info:
...
+    | 7.  The refresh token can be used to get a new OAuth `access_token` for another 2 hours by contacting the authentication server again with the `client_id`, `client_secret`, and `refresh_token`. | |
```
Comment (Dorthu) [inline openapi.yaml]: You also get back a new refresh token, allowing you to continue to keep the session alive.

File: openapi.yaml [hunk 5]
```diff
@@ -47,33 +47,45 @@ info:
...
+
+      Once the User has logged into Linode and you have received an exchange code,
```
Comment (Dorthu) [inline openapi.yaml]: Might want to qualify that this only applies to the private workflow.

File: openapi.yaml [hunk 6]
```diff
@@ -47,33 +47,45 @@ info:
...
-    | PARAMETER | DESCRIPTION |
-    |-----------|-------------|
-    | client_id | Your app's client ID |
-    | client_secret | Your app's client secret |
-    | code | The code you just received from the redirect |
+      | PARAMETER | DESCRIPTION | FIND OR UPDATE THIS INFORMATION |
+      |-----------|-------------|---------------------------------|
+      | client_id | Your app's client ID. | [List your OAuth Clients](https://developers.linode.com/api/v4/account-oauth-clients). |
+      | client_secret | Your app's client secret. | [View your OAuth Client](https://developers.linode.com/api/v4/account-oauth-clients-client-id). [Reset Your OAuth Client Secret](https://developers.linode.com/api/v4/account-oauth-clients-client-id-reset-secret/#post). |
+      | code | The code you just received from the redirect | [View your OAuth Client](https://developers.linode.com/api/v4/account-oauth-clients-client-id). [Reset Your OAuth Client Secret](https://developers.linode.com/api/v4/account-oauth-clients-client-id-reset-secret/#post). |
```
Comment (Dorthu) [inline openapi.yaml]: These two links don't apply to the exchange code, and there's no way to retrieve it if you don't capture it in the redirect.

File: openapi.yaml (hunk 1 above)
Comment (hzoppetti) [reply openapi.yaml]: fixed

File: openapi.yaml (hunk 2 above)
Comment (hzoppetti) [reply openapi.yaml]: changed to include both options

File: openapi.yaml (hunk 3 above)
Comment (hzoppetti) [reply openapi.yaml]: updated this to be more clear

File: openapi.yaml (hunk 4 above)
Comment (hzoppetti) [reply openapi.yaml]: updated to be more clear

File: openapi.yaml (hunk 5 above)
Comment (hzoppetti) [reply openapi.yaml]: added a note to say this extends the private workflow

File: openapi.yaml (hunk 6 above)
Comment (hzoppetti) [reply openapi.yaml]: removed the link column because it was confusing

File: openapi.yaml [hunk 7]
```diff
@@ -47,19 +47,43 @@ info:
...
+    | 4.  The authorization server redirects the user to the specificed redirect URL with a temporary authorization `code` (exchange code) in the URL. | 4.  The authorization server redirects the user back to your application with an OAuth `access_token` embedded in the redirect URL's hash. This is temporary and expires in 2 hours without a `refresh_token`. |
+    | 5.  The application issues a POST request (*see below*) to the authentication server with the exchange code, `client_id`, and the client application's `client_secret`. | |
+    | 6.  The authentication server responds to the client application with a new OAuth `access_token` and `refresh_token`. The `access_token` is set to expire in 2 hours. | |
+    | 7.  The `refresh_token` can be used to get a new OAuth `access_token`, good for another 2 hours, and a new `refresh_token`, which you can use to extend the session again by this same method, by contacting the authentication server again with the `client_id`, `client_secret`, and `refresh_token`. | |
```
Comment (nmelehan) [inline openapi.yaml]: it seems like we might be using `authentication server` and `authorization server` somewhat interchangeably here, which could maybe become confusing? I've read that OAuth makes distinctions between authentication and authorization:
https://stackoverflow.com/questions/33702826/oauth-authorization-vs-authentication

Maybe we can decide to just use one of these? I'm not so great with OAuth, so if a @Dorthu or @jnschaeffer could offer some insight too, that'd be great

Comment (nmelehan) [inline openapi.yaml]: I think step 7 is a bit confusing. I thought about splitting it into two sentences, but I'm not actually sure about what a refresh token is capable of doing. As written, it sounds like a refresh token can refresh itself... which doesn't seem right... but maybe it is?

File: openapi.yaml [hunk 8]
```diff
@@ -47,19 +47,43 @@ info:
...
+    | 3.  The user logs into the authorization server with their username and password. | 3.  The user logs into the authorization server with their username and password. |
+    | 4.  The authorization server redirects the user to the specificed redirect URL with a temporary authorization `code` (exchange code) in the URL. | 4.  The authorization server redirects the user back to your application with an OAuth `access_token` embedded in the redirect URL's hash. This is temporary and expires in 2 hours without a `refresh_token`. |
+    | 5.  The application issues a POST request (*see below*) to the authentication server with the exchange code, `client_id`, and the client application's `client_secret`. | |
+    | 6.  The authentication server responds to the client application with a new OAuth `access_token` and `refresh_token`. The `access_token` is set to expire in 2 hours. | |
```
Comment (nmelehan) [inline openapi.yaml]: Just want to confirm with @dorthu or @jnschaeffer that it's correct that the token endpoint will return both an access token and a refresh token for private/confidential oauth clients (described in step 6 of the private workflow)

File: openapi.yaml [hunk 9]
```diff
@@ -47,19 +47,43 @@ info:
...
+    | 1.  The User visits the application's website and is directed to login with Linode. | 1.  The User visits the application's website and is directed to login with Linode. |
+    | 2.  Your application then redirects the user to Linode's [authentication server](https://login.linode.com) with the client application's `client_id` and requested OAuth `scope`, which should appear in the URL of the login page. | 2.  Your application then redirects the user to Linode's [authentication server](https://login.linode.com) with the client application's `client_id` and requested OAuth `scope`, which should appear in the URL of the login page. |
+    | 3.  The user logs into the authorization server with their username and password. | 3.  The user logs into the authorization server with their username and password. |
+    | 4.  The authorization server redirects the user to the specificed redirect URL with a temporary authorization `code` (exchange code) in the URL. | 4.  The authorization server redirects the user back to your application with an OAuth `access_token` embedded in the redirect URL's hash. This is temporary and expires in 2 hours without a `refresh_token`. |
```
Comment (nmelehan) [inline openapi.yaml]: Can we more explicitly state that a refresh token is *not* returned by the login server (step 4 of the public workflow)

File: openapi.yaml (hunk 7 above)
Comment (hzoppetti) [reply openapi.yaml]: I agree - this is confusing, our doc also uses both of these terms in this way, if they are different that's fine. If this is all the same, authentication server, we should update the doc as well.

File: openapi.yaml (hunk 9 above)
Comment (hzoppetti) [reply openapi.yaml]: updated this to be more explicit

File: openapi.yaml (hunk 7 above)
Comment (hzoppetti) [reply openapi.yaml]: Updated this to be more clear.

Comment (nmelehan) [reply openapi.yaml]: @hzoppetti and I talked to @jnschaeffer, resolved this in subsequent commits

**Pull request merged**